### DISCUSSION TITLE
Simple way to evaluate ratio plot in uproot?

### DISCUSSION THREAD

**Discussion by qibin2020** [scikit-hep/uproot5#642] [Q&A] [answered]
Hi I am very new to uproot from ROOT. I was wondering is there some recommended way to do histogram division in uproot?

  As in my previous ROOT workflow we frequently calculate ratio histograms by TH1.Divide(TH1), and store them for later usage(e.g. crazy ratio of ratio) or plotting. I am not sure how to adapt this workflow in uproot since I haven't find equivalent method in any hist or boost-histogram Hist object. I found the hist.plot_ratio is useful but not easy to generalized to, e.g. plot multiple ratios. 

  So does anyone have some ideas on this or we may still need to import ROOT to handle the histogram division operation and finally convert to uproot-hist?

  I am using uproot4.2.0 shipped with scikit-hep  [v3.6.0](https://github.com/scikit-hep/scikit-hep/releases/tag/v3.6.0).

  (Pls correct me if asked at a wrong place and many thanks)

**Comment by MoAly98** [comment, marked as answer]
Hi Qibin! I am not an expert, but as far as I understand the division of histograms in `boost_histogram` is not fully complete yet. AFAIK the division of weighted histograms is not done yet (see [boost-histogrma issue](https://github.com/boostorg/histogram/issues/352)).  Possibly one way around this is to define your own `divide` function to divide the histogram values as `numpy` arrays then compute the error using the [ratio_uncertainty method from scikit-hep/Hist ](https://hist.readthedocs.io/en/latest/reference/hist.intervals.html?highlight=ratio_uncertainty#hist.intervals.ratio_uncertainty) ?  You can then assign to the results to a boost-histogram/ Hist object. 

I am sure there are better ideas from the experts here, but maybe that's one way :)

**Reply by qibin2020** [reply]
Thanks a lot! I think this is a good workaround and I will have a try. (Also thanks for the suggestion of `ratio_uncertainty`, I realize the uncertainty of ratio is not trivial so this might be a challenge for the generalization)

**Reply by jpivarski** [reply]
I think @MoAly98's suggestion is the best one: get the NumPy arrays and do the division manually. Histogram division is tricky because it will assume some error propagation, and usually the wrong one. Are the bins of the numerator statistically independent of the bins of the denominator (as is often the case with histogram addition)? Or is the numerator a _subset_ of the denominator (i.e. highly dependent, but a common use-case: efficiency plots)?

Although histogram packages can implement these, it can be hard to be sure that you're communicating with the histogram package on what the error bars for the histogram mean, and since this is vital to the final result, doing it explicitly (with NumPy arrays) is prudent for self-documentation, not just a work-around. Maybe if the histogram package has very well-named methods, like `divide_as_independent` or `divide_as_subset`, the intention would be clear, but if it's just an overloaded `/` operator, that's dangerous. Even if it works, a reader of the code might not realize your intent.

**Reply by qibin2020** [reply]
Yes I agree. The `/` operation is not well-defined for the histogram itself without the knowledge of the statistical meaning of the deno/nom.

Besides I just found the codes of hist.plot_hist() have `ratio_uncertainty` and options for multiple uncertainty cases even the `pull`. So maybe inheriting this func and extending the functionality (like multiple ratio plotting) is also a convenient workflow for analysis.

Marked @MoAly98 's as answer and thanks for the helpful discussion!

**Reply by jpivarski** [reply]
(Thanks for marking answers!)
